### PR TITLE
Speed up column index retrieval

### PR DESCRIPTION
**Pull request by wsmigaj**
This PR makes the `TODATableIterator::updateMetaData()` function more efficient by
* performing the search for each column only once (by calling just `MetaData::columnIndex()`) rather than twice (with an extra call to `MetaData::hasColumn()`)
* avoiding repeated construction of a temporary string representing the column name with a `@` suffix
* removing an unnecessary call to `std::string::find()`.

On my system, this reduces the execution time of a query operating on a file containing 1,292,800 rows and 212 columns and producing an empty results set (and therefore effectively only decoding the file), i.e.
```
odc sql "select * where varno == -1" -i atms6K.odb -o /dev/null -f odb
```
by more than 25%: from 2.0 s to 1.4 s.

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](http://claassistant.ecmwf.int:80/ecmwf/odc?pullRequest=5) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](http://claassistant.ecmwf.int:80/ecmwf/odc?pullRequest=5) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](http://claassistant.ecmwf.int:80/check/ecmwf/odc?pullRequest=5) it.</sub>

**Comment by wsmigaj** [comment]
@simondsmart Would you mind having a look at this PR when you have time (or suggesting another reviewer)?

**Comment by simondsmart** [comment]
@wsmigaj - many thanks for this. This looks like a good change, and I have only a few comments, which I will put inline.

Could I ask for a copy of the data (or a reduced set) that is exhibiting the performance improvement - it is larger than I would expect! We should (certainly) not be checking column existence/indices sufficiently frequently to impact runtime by 25%!

Thanks for your input!

File: src/odc/core/MetaData.cc [hunk 1]
```diff
@@ -301,6 +304,18 @@ MetaData& MetaData::addBitfield(const std::string& name, const eckit::sql::Bitfi
 
 //----------------------------------------------------------------------------------------------------------------------
 
+bool columnNameMatches(const std::string &fullColumnName,
```
Comment (simondsmart) [inline src/odc/core/MetaData.cc:307]: This is much improved. Thanks!

File: src/odc/core/MetaData.cc [hunk 2]
```diff
@@ -87,26 +88,28 @@ Column* MetaData::columnByName(const std::string& name) const
...
-		if (at(i)->name() == name || at(i)->name().find(name + "@") == 0)
+		if (columnNameMatches(at(i)->name(), name))
 			return true;
 	return false;
 }
 
 size_t MetaData::columnIndex(const std::string& name) const
 {
-	std::vector<size_t> indices;
+	const size_t notFound = static_cast<size_t>(-1);
```
Comment (simondsmart) [inline src/odc/core/MetaData.cc:98]: I would prefer not abusing signed/unsigned values here.
Either use a signed integer within the routine (and cast to size_t after the check that it is >= 0), or use a boolean 'found' for the check.

File: src/odc/core/MetaData.cc [hunk 3]
```diff
@@ -87,26 +88,28 @@ Column* MetaData::columnByName(const std::string& name) const
...
-		if (at(i)->name() == name || at(i)->name().find(name + "@") == 0)
-			indices.push_back(i);
-
-	if (indices.size() > 1)
-		throw eckit::UserError(std::string("Ambiguous column name: '") + name + "'");
+		if (columnNameMatches(at(i)->name(), name)) {
+			if (index == notFound)
+				index = i;
+			else
+				throw eckit::UserError(std::string("Ambiguous column name: '") + name + "'");
```
Comment (simondsmart) [inline src/odc/core/MetaData.cc:106]: Please create a derived exception type (see comment below). Something along the lines of:
class AmbiguousColumnException : public UserError {
    AmbiguousColumnException(const std::string& columnName) { ...}
    ....
};
This lets us do the exception wrapping that you do below more cleanly.

File: src/odc/core/MetaData.cc [hunk 4]
```diff
@@ -87,26 +88,28 @@ Column* MetaData::columnByName(const std::string& name) const
...
+		}
 
-	if (indices.size() == 0)
+	if (index == notFound)
 		throw eckit::UserError(std::string("Column '") + name + "' not found.");
```
Comment (simondsmart) [inline src/odc/core/MetaData.cc:110]: Same here.

File: src/odc/sql/TODATableIterator.cc [hunk 5]
```diff
@@ -23,6 +24,33 @@ namespace sql {
...
+/// Return the index of column `columnName` in metadata. If not found or ambiguous,
+/// throw an exception.
+size_t columnIndex(const std::string& columnName, const core::MetaData &md)
+{
+    size_t idx;
+    try {
+        idx = md.columnIndex(columnName);
+    } catch (const eckit::UserError &e) {
+        // Make some error messages more precise
+        if (eckit::StringTools::startsWith(e.what(), "Ambiguous column name:"))
```
Comment (simondsmart) [inline src/odc/sql/TODATableIterator.cc:38]: This is something of a code smell.
By preference, I would write the inner exceptions such that they don't need to be wrapped.
But if the message cannot be written sufficiently generically, such that you really need to put the "SQL request" text in here, please define derived exception types, and then catch them specifically. Looking at the string contents is something of a smell, and is very much more brittle.

**Review by simondsmart — CHANGES_REQUESTED** [review]

**Comment by wsmigaj** [comment]
Thank you very much for the rapid review and your comments. I agree it is much better to use more specific exception types -- I've modified the code as you suggested.

What is the best way to share the file with you (I would prefer not to post it publicly on GitHub)? It is relatively large (107 MB).

**Comment by simondsmart** [comment]
Please drop it into my dropbox received folder, using this file request: https://www.dropbox.com/request/qsvtHLikyKl6wIThaBcV

**Comment by wsmigaj** [comment]
@simondsmart I've uploaded the file to your DropBox folder (you'll probably have received a notification from DropBox -- commenting here just in case you won't).

**Comment by wsmigaj** [comment]
@simondsmart This is just a gentle reminder about this PR. Have you perhaps had a chance to benchmark the code using the provided ODB file? Are you happy with the way your comments have been addressed?

**Review by simondsmart — APPROVED** [review]

**Comment by simondsmart** [comment]
Thanks @wsmigaj, much appreciated.
I'm happy with the PR - much improved over what was there before. I've been a bit overtaken with some recent personnel changes here, but hopefully I will get the chance to do the benchmarking soon! Either way, the code is improved :-).

**Comment by DJDavies2** [comment]
Is there any chance of getting this on a tag at some point?

**Pull request merged**